### PR TITLE
[TextSelectionControls] move the tap gesture callback into _TextSelectionHandlePainter, remove `_TransparentTapGestureRecognizer`.

### LINKED ISSUE #39376: Taps on TextField prefix/suffix
I'm experiencing weird issues with the latest stable version of Flutter: 1.7.8+hotfix.4.
I have a simple textfield widget that has an `onTap()` action. The textfield widget also has a prefix icon which has its own `onTap()` action. When I press the prefix icon, it completes everything inside the prefix `onTap()` but it also then goes and completes everything in the textfield `onTap()`.

I've been able to replicate it with the below code. Both onTap methods print outputs to the console. When the prefix icon is clicked, it is outputting the icon onTap as well as the textfield onTap actions.

**Replicated in below code**
```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  void _incrementCounter() {
    setState(() {
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              color: Colors.red,
              child: TextField(
                onTap: () {
                  print('Hello this is a print out');
                },
                decoration: InputDecoration(
                  hintText: 'Hint text',
                  border: InputBorder.none,
                  contentPadding: EdgeInsets.only(left: 25.0, top: 16.0),
                  prefixIcon: IconButton(
                    icon: Icon(Icons.hotel, color: Colors.green, size: 40.0,),
                    onPressed: () {
                      print("Hi this is a second print out");
                    },
                  ),
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}
```
**Flutter Doctor**
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.7.8+hotfix.4, on Mac OS X 10.14.6 18G87, locale en-NZ)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
[✓] iOS tools - develop for iOS devices
[✓] Android Studio (version 3.4)
[✓] VS Code (version 1.37.1)
[✓] Connected device (1 available)

• No issues found!
```
I have switched back to the master channel in the mean time to get around this issue but would be good to have it fixed. 
Thanks.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
For iOS/Android behaviors see https://github.com/flutter/flutter/issues/39376#issuecomment-848406415

Fixes https://github.com/flutter/flutter/issues/39376.

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/material/text_field_test.dart [hunk 1]
```diff
@@ -489,7 +489,7 @@ void main() {
     await tester.tap(find.byKey(icon));
     await tester.pump();
     expect(controller.text, '');
-    expect(controller.selection, const TextSelection.collapsed(offset: 0, affinity: TextAffinity.upstream));
```
Comment (chunhtai) [inline packages/flutter/test/material/text_field_test.dart:492]: Why this change?

File: packages/flutter/test/material/text_field_test.dart [hunk 2]
```diff
@@ -9719,4 +9719,44 @@ void main() {
...
+
+    int textFieldTapCount = 0;
+    int prefixTapCount = 0;
+    int suffixTapCount = 0;
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: TextField(
+            //focusNode: focusNode,
```
Comment (chunhtai) [inline packages/flutter/test/material/text_field_test.dart:9734]: remove?

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 3]
```diff
@@ -401,11 +401,19 @@ class TextSelectionOverlay {
   final DragStartBehavior dragStartBehavior;
 
   /// {@template flutter.widgets.TextSelectionOverlay.onSelectionHandleTapped}
-  /// A callback that's invoked when a selection handle is tapped.
-  ///
-  /// Both regular taps and long presses invoke this callback, but a drag
-  /// gesture won't.
+  /// A callback that's optionally invoked when a selection handle is tapped.
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:404]:
```suggestion
  /// A callback that's optionally invoked based on platforms when a selection handle is tapped.
```

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 4]
```diff
@@ -401,11 +401,19 @@ class TextSelectionOverlay {
...
+  /// The [TextSelectionControls.buildHandle] implementation the text field
+  /// uses decides where the the handle's tap "hotspot" is, or whether the
+  /// selection handle supports tap gestures at all. For instance,
+  /// [MaterialTextSelectionControls] calls [onSelectionHandleTapped] when the
+  /// selection handle's "knob" is tapped, while
+  /// [CupertinoTextSelectionControls] builds a handle that's not sufficiently
+  /// large for tapping (as it's not meant to be tapped) so it does not call
+  /// [onSelectionHandleTapped] even when tapped.
   /// {@endtemplate}
+  // See https://github.com/flutter/flutter/issues/39376#issuecomment-848406415
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:415]: I think we avoid pointing to a github link in API doc

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 5]
```diff
@@ -114,7 +114,7 @@ abstract class TextSelectionControls {
   ///
   /// The top left corner of this widget is positioned at the bottom of the
   /// selection position.
-  Widget buildHandle(BuildContext context, TextSelectionHandleType type, double textLineHeight);
+  Widget buildHandle(BuildContext context, TextSelectionHandleType type, double textLineHeight, VoidCallback? onTap);
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:117]: document what onTap is

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 6]
```diff
@@ -855,18 +860,18 @@ class _TextSelectionHandleOverlayState
...
-            onTap: _handleTap,
             child: Padding(
               padding: EdgeInsets.only(
                 left: padding.left,
                 top: padding.top,
                 right: padding.right,
                 bottom: padding.bottom,
               ),
-              child: widget.selectionControls!.buildHandle(
+              child: widget.selectionControls.buildHandle(
```
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/text_selection.dart:870]: is it possible we wraps the widget from the buildhandle with the gesturedetector here? It feels weird that we ask the widget to build the gestureDetector.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:409]: I think ultimately this depends on what kind of selection control you're using, so it's not really platform-dependant. Even in `TextField`'s implementation you can override `textSelectionControl` so it's still up to the developer.

Anyways IMO we should deprecate this callback and let the `TextSelectionControl` implementation decides how the handles should be built and what interactions it should support. Currently it does not have enough information (as it doesn't have access to the `EditableTextState` that builds the selection overlay).

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:420]: The last 2 lines are prefixed by "//".

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 6 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:875]: It's the selection control of the user's choice that builds the handle right? The selection control decides the style of the handles and the types of interactions it suppoorts.

File: packages/flutter/test/material/text_field_test.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/material/text_field_test.dart:492]: I think before the change the tap event actually altered the selection. Now it's only hitting the button and not the text field so that should be expected.

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 6 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:875]: It's a bit subtle that `onTap` should only work on the "knob" , while drag gestures should work on both the knob and the rectangle attached to the knob.

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/text_selection.dart:875]: That is an unfortunate how onSelectionHandleTapped behave differently based on the implementation of TextSelectionControls. 

I am ok with this approach if we can deprecate the onSelectionHandleTapped at some point....

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 7]
```diff
@@ -110,11 +110,16 @@ class ToolbarItemsParentData extends ContainerBoxParentData<RenderBox> {
...
   ///
   /// The top left corner of this widget is positioned at the bottom of the
   /// selection position.
-  Widget buildHandle(BuildContext context, TextSelectionHandleType type, double textLineHeight);
+  ///
+  /// The supplied [onTap] should be invoked when the handle is tapped, if such
+  /// interaction is allowed. As a counterexample, the default selection handle
+  /// on iOS [cupertinoTextSelectionControls] does not call [onTap] at all,
+  /// since its handles are not meant to be tapped.
+  Widget buildHandle(BuildContext context, TextSelectionHandleType type, double textLineHeight, VoidCallback? onTap);
```
Comment (creativecreatorormaybenot) [inline packages/flutter/lib/src/widgets/text_selection.dart:122]: @LongCatIsLooong This is a **breaking change** 😓 How are we supposed to deal with packages that should work with multiple versions of Flutter that call `buildHandle`.

https://pub.dev/packages/flutter_math_fork is such a package and now it cannot work.

The change should have been `[VoidCallback? onTap]`, otherwise it is breaking. Should I make a PR and we cherry pick a hot fix for the dev channel?

Comment (creativecreatorormaybenot) [reply packages/flutter/lib/src/widgets/text_selection.dart:122]: So `beta` and `stable` have 3 positional parameters and `dev` has 4 positional parameters...

**Comment by creativecreatorormaybenot** [comment]
@LongCatIsLooong This is how I tried to fix it: https://github.com/simpleclub-extended/flutter_math_fork/commit/7c13f1e54f2e922ef246a85f52e2fd0b8c9074cd#diff-0cee011dc67be19dc0c38a0f8a915554273b9f7f1b4c38402db2112d16a7d596R175

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 7 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/text_selection.dart:122]: Yeah the proposed change sounds good to me. Sorry for the breakage!

Per the [breaking change policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#1-determine-if-your-change-is-a-breaking-change) this is not a breaking change so I'd recommend contributing tests to `flutter/tests` to prevent future breakage.
I'm not sure we have a process for cherry-picking a commit to the dev channel. See https://github.com/flutter/flutter/wiki/Flutter-Cherrypick-Process. Maybe we should add that dev build to https://github.com/flutter/flutter/wiki/Bad-Builds.

**Comment by justinmc** [comment]
Thanks for fixing this @LongCatIsLooong!  Good straightforward idea.  This bug has been coming up over and over for 2 years.
